src/peft/tuners/gralora/layer.py

```diff
@@ -0,0 +1,427 @@
+        """
+        Move the adapter of the given name to the device of the base layer.
+        """
+        from peft.tuners._buffer_dict import BufferDict
+
+        if device is None:
+            # check weight and qweight (for GPTQ)
+            for weight_name in ("weight", "qweight"):
+                weight = getattr(self.get_base_layer(), weight_name, None)
+                if weight is not None:
+                    device = weight.device
+                    dtype = weight.dtype
+                    break
+            else:
+                # no break encountered: could not determine the device
+                return
+
+        # loop through all potential adapter layers and move them to the device of the base layer; be careful to only
+        # move this specific adapter to the device, as the other adapters could be on different devices
+        # see #1639
+        for adapter_layer_name in self.adapter_layer_names + self.other_param_names:
+            adapter_layer = getattr(self, adapter_layer_name, None)
+            if not isinstance(adapter_layer, (nn.ModuleDict, nn.ParameterDict, BufferDict)):
+                continue
+            if adapter_name not in adapter_layer:
+                continue
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                adapter_layer[adapter_name] = adapter_layer[adapter_name].to(device, dtype=dtype)
+            else:
+                adapter_layer[adapter_name] = adapter_layer[adapter_name].to(device)
+
+    @property
+    def merged(self) -> bool:
+        return bool(self.merged_adapters)
+
+    @property
+    def bias(self) -> torch.Tensor:
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            return base_layer.bias
+        elif isinstance(base_layer, Conv1D):
+            return base_layer.bias
+        else:
+            return None
+
+    def update_layer(
+        self,
+        adapter_name,
+        module_name,
+        r,
+        gralora_alpha,
+        gralora_dropout,
+        gralora_k: int = 2,
+        hybrid_r: int = 0,
+        init_weights: bool = True,
+    ):
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+
+        self.r[adapter_name] = r
+        self.gralora_alpha[adapter_name] = gralora_alpha
+        self.gralora_k[adapter_name] = gralora_k
+        self.hybrid_r[adapter_name] = hybrid_r
+
+        if gralora_dropout > 0.0:
+            gralora_dropout_layer = nn.Dropout(p=gralora_dropout)
+        else:
+            gralora_dropout_layer = nn.Identity()
+
+        self.gralora_dropout.update(nn.ModuleDict({adapter_name: gralora_dropout_layer}))
+
+        # Actual trainable parameters
+        subblock_in_features = self.in_features // gralora_k
+        subblock_out_features = self.out_features // gralora_k
+
+        gralora_r = r - hybrid_r  # gralora_r is the rank allocated to gralora method
```

Comment:
Just wondering if it's a good idea to subtract `hybrid_r` from `r`. Would it not be easier to handle if we just leave `r` as it is? I know it would be different from how `hybrid_r` is handled in the original repo, but I find it more intuitive.